src/proxy/h2.rs

```diff
@@ -138,6 +140,53 @@ impl Drop for DropCounter {
     }
 }
 
+// We can't directly implement tokio::io::{AsyncRead, AsyncWrite} for H2Stream because
+// then the specific implementation will conflict with the generic one.
+impl TokioH2Stream {
+    pub fn new(stream: H2Stream) -> Self {
+        Self(stream)
+    }
+}
+
+impl tokio::io::AsyncRead for TokioH2Stream {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut tokio::io::ReadBuf<'_>,
+    ) -> Poll<std::io::Result<()>> {
+        let pinned = std::pin::Pin::new(&mut self.0.read);
+        copy::ResizeBufRead::poll_bytes(pinned, cx).map_ok(|bytes| buf.put(bytes))
```

Comment:
This seems like there is no protection against ReadBuf's rules: `Panics if self.remaining() is less than buf.len().`

Reply:
Can we plausible write a test around this like we have in copy.rs wrapping in `WeirdIO`? It detects these types of bugs

Reply:
Yes, I'm holding back on unit tests until there is some agreement on the implementation